Кечин Валерий, Вариант 6 (Tar)

kechinvv opened the pull request.

kechinvv committed:
First Version

Br0mm commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+public class Tar {
+    public static void main(String[] args) throws IOException {
+        Path out = Paths.get(args[args.length - 1]);
+        if (args[args.length - 2].equals("-out")) {
```
Флаг -out не обязательно будет находиться в конце. Он может быть указан и до входных файлов

kechinvv replied:
Исправил

Br0mm commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+                    fin.write("\n" + line);
+                }
+            }
+            fin.close();
```
Закроется ли файл, если в коде произойдет исключение?

kechinvv replied:
Исправил

Br0mm commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+            }
+            for (int i = 0; i < args.length - 2; i++) {
+                for (String line : readAllLines(Paths.get(args[i]))) {
+                    fin.write("\n" + line);
```
Есть переносы строк `\n,` а есть `\r\n`. Если файлы с переносом строк `\r\n` запаковать, потом распаковать то будут ли это те же самые файлы?

kechinvv replied:
Проверил, будут в обоих ситуациях, в новых тестах в file1.txt добавил строку с таким переносом

Br0mm commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+                String[] a = fin.get(i).split(" ");
+                BufferedWriter f = newBufferedWriter(Paths.get(a[0]), WRITE, CREATE, TRUNCATE_EXISTING);
+                for (int l = start; l<=start+Integer.parseInt(a[1])-1;l++ ) {
+                    f.write(fin.get(l)+"\n");
```
Аналогично с переносами `\n` и `\r\n`.

Br0mm commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+                    f.write(fin.get(l)+"\n");
+                }
+                start+=Integer.parseInt(a[1]);
+                f.close();
```
Аналогично с закрытием файла при исключении.

Br0mm commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+            }
+        }
+    }
+}
```
Упаковку файлов и распаковку файлов лучше вынести в отдельные классы. 

Так же лучше сделать класс для разбора аргументов командной строки.

h31 replied:
Упаковка и распаковка файлов могут быть в одном классе, ничего плохого в этом вижу. Разбор аргументов действительно стоит вынести, как минимум для удобства тестирования. Парсер аргументов - это, фактически, контроллер из MVC.

Br0mm commented on src/test/java/TestTar.java:
```diff
@@ -0,0 +1,67 @@
+
+public class TestTar {
+    @Test
+    public void test() {
```
Необходимо структуризировать тесты:
- Разделить тесты на тестирование "чёрным" и "белым" ящиком
- Разбить тесты на несколько методов, в котором будет проверяться своя часть программы, а не писать одним методом.
- Добавь проверку, что ты запаковываешь файлы, затем этот архив и ещё файлы ещё раз запаковываешь, а потом всё это распаковываешь и изначальное должно сходиться с полученным.
- Не хватает теста парсера аргументов командной строки.

h31 replied:
Да, сейчас основная проблема - это тесты. Очень длинный тест, в котором проверяется всё. Такой тест сложно читать, сложно при необходимости менять. Если тест окажется провален, то будет совершенно непонятно, что именно сломалось. Сама идея протестировать всё сразу имеет смысл, это называется системным тестированием, но нужно разобраться с длиной и удобочитаемостью теста, плюс сделать тесты для отдельных компонентов приложения - распаковщика, упаковщика, парсера аргументов. Не забыть проверить некорректные сценарии использования и вывод ошибок.

h31 replied:
> Разделить тесты на тестирование "чёрным" и "белым" ящиком

Справедливости ради, black box testing и white box testing - это просто две стратегии тестирования. Стратегии помогают, когда мы хотим написать качественные тесты, но не понимаем, как лучше это сделать. У каждой из этих двух стратегий есть формальные критерии, но в реальной жизни зачастую получаются только grey box тесты той или иной степени. Т.е. это не жесткая классификация.

В первом задании white box testing представлял особый интерес для того, чтобы проверить внешнее поведение нашего модуля, поэтому я иногда просил вынести подобные тесты в отдельный класс. И хотя разделить тесты - идея неплохая, не могу сказать, что это обязательное требование. В данном задании скорее актуально разделение по уровням - модульные, интеграционные, системные тесты.

kechinvv replied:
Разделил тесты, добавил тест для парсера, также разделил тесты на проверку исключений.

h31 commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+    public static void main(String[] args) throws IOException {
+        Path out = Paths.get(args[args.length - 1]);
+        if (args[args.length - 2].equals("-out")) {
+            BufferedWriter fin = newBufferedWriter(out, WRITE, CREATE, TRUNCATE_EXISTING);
```
В чем смысл этих флагов?

kechinvv replied:
В nio.file write необходим для начала записи, create для создания файла если таковой отсутствует, truncate стирает все из файла при его перезаписи.

h31 commented on src/main/java/Tar.java:
```diff
@@ -0,0 +1,43 @@
+                fin.write("\n" + args[i] + " " + readAllLines(Paths.get(args[i])).size());
+            }
+            for (int i = 0; i < args.length - 2; i++) {
+                for (String line : readAllLines(Paths.get(args[i]))) {
```
Дважды считываете файл. Сделайте так, чтобы это происходило один раз.

h31 commented on src/test/java/TestTar.java:
```diff
@@ -0,0 +1,67 @@
+            Tar.main(args);
+            assertEquals(readAllLines(Paths.get("output.txt")), readAllLines(Paths.get("outputnew.txt")));
+        } catch (IOException e) {
+            e.printStackTrace();
```
Разве появление исключения не должно привести к провалу теста? Сейчас мы по факту игнорируем исключения, просто выводим сообщение на экран.

kechinvv committed:
Second Version

kechinvv committed:
Version 2.1
Added:
1. Descriptions of exceptions for users
2. Warning of probable user's mistake
